### ISSUE TITLE
Best way to convert field(s) to and from CF-compliant `xarray.Dataset`

### ISSUE THREAD

**Issue by sadielbartholomew** [NCAS-CMS/cf-python#773]
A common way to work with weather and climate datasets is of course the [xarray](https://docs.xarray.dev/en/stable/user-guide/weather-climate.html#weather-and-climate-data) library, so it would be nice to have a means to convert to and from an [`xarray.Dataset`](https://docs.xarray.dev/en/stable/generated/xarray.Dataset.html#xarray.Dataset) if it has sufficient and conformant-enough CF metadata and suitable structure, for which [the `xarray.decode_cf` function](https://docs.xarray.dev/en/stable/generated/xarray.decode_cf.html#xarray.decode_cf) may be able to assist on the xarray-side.

Ideally we can have a method for such a conversion and/or at least a recipe in the documentation showing how this can be done, for both a trivially-convertible case and for a case of a field and/or `Dataset` that needs some tweaks to get the right structure before the conversion can be made.

Note in #706 we considered and supported input of xarray arrays, so that helps towards this.

**Comment by bouweandela** [comment]
Maybe it could be interesting to explore if you could add a `to_cfpython` and `from_cfpython` function to the [ncdata](https://ncdata.readthedocs.io/en/latest/userdocs/getting_started/introduction.html#working-with-netcdf-files) package developed by @pp-mo. That would allow users to seamlessly switch between all of the most commonly used Python weather and climate data analysis software packages, even within a single analysis.

A really nice feature of the `ncdata` package is that it provides a clone of the NetCDF4 Dataset API and functions to convert Xarray Datasets to/from it. This allows users who have some data in Xarray that does not comply with the CF conventions (e.g. loaded from Zarr using a 3rd party object store or a NetCDF file in a central location on an HPC system), to convert their Xarray Dataset to an in-memory NetCDF Dataset, fix the NetCDF attributes to make the data CF compliant, and then interpret that with a package that works with the CF conventions. This is convenient for users, as it allows loading any data supported by Xarray in a computationally efficient way while still benefitting from the advantages that CF-aware packages like cf-python and Iris bring. It is also convenient from a software maintenance point of view, as it removes the need for custom logic to support all the possible ways in which datasets can be non-compliant with the CF conventions.

**Comment by davidhassell** [comment]
OK - we're half way there. The next version (v3.20.0 - end of March 2026) will have complete conversion-to-xarray capability (I'll link an issue soon).

The will take the form of the following methods:
- `cf.Field.to_xarray`
- `cf.FieldList.to_xarray`
- `cf.Domain.to_xarray`
- `cf.DomainList.to_xarray`

As well as the ability to create an xarray dataset in memory directly from `cf.write`: `ds = cf.write(..., fmt='XARRAY')`

The newly created xarray dataset in all five cases is either an `xarray.Dataset` instance, or if there are implied netCDF groups, an `xarray.DataTree` instance (if desired).

In all cases the conversion is "lossless, copy-free and lazy-preserving" (to pinch the nice phrase from ncdata :)).

The conversion is as good as cf-python's ability to write a Field or Domain to disk - i.e. it is 100% CF -compliant and suffers 0% CF information loss. This is because the conversion is actually done by `cf.write` using `xarray` as its netCDF backend (as opposed to the more usual options that write to disk: `netCDF4`, `h5netcdf`, `zarr`, etc.) - so all of the CF-data-model-to-encoded-dataset logic defined in `cf.write` can be brought to bear "for free".

----

Conversion from xarray to cf-python Fields or Domains is also on the way, but not until in v3.21.0
